bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java

```diff
@@ -21,106 +21,234 @@
+            default:
+                return ValueMap.EMPTY;
+        }
+    }
+
+    ValueMap getValueMapFromBindings(final SlingBindings bindings,
+                                     final SharedComponentProperties.ValueTypes valueType,
+                                     final Resource resource,
+                                     final String rootPagePath) {
+        // if the merged path in bindings matches the resource path, just assume that the merged properties in
+        // bindings are sufficient
+        if (valueType == SharedComponentProperties.ValueTypes.MERGED
+                && resource.getPath().equals(bindings.get(SharedComponentProperties.MERGED_PROPERTIES_PATH))) {
+            return Optional.ofNullable((ValueMap) bindings.get(SharedComponentProperties.MERGED_PROPERTIES))
+                    .orElse(ValueMap.EMPTY);
+        }
+
+        // next check that the root page path matches the shared properties page path in bindings
+        // this might not match if a request wrapper was constructed for a non-child resource of the original request
+        // but which was not subjected to its own script bindings phase
+        if (!rootPagePath.equals(bindings.get(SharedComponentProperties.SHARED_PROPERTIES_PAGE_PATH))) {
+            // return null to indicate that the Injector should construct the value map
+            return null;
+        }
+
+        final ValueMap globalVmBound = (ValueMap) bindings.get(SharedComponentProperties.GLOBAL_PROPERTIES);
+        // if GLOBAL is requested, just return it from bindings
+        if (valueType == SharedComponentProperties.ValueTypes.GLOBAL) {
+            return Optional.ofNullable(globalVmBound).orElse(ValueMap.EMPTY);
+        }
+
+        final String sharedPropertiesPath = sharedComponentProperties.getSharedPropertiesPath(resource);
+        // it is possible for getSharedPropertiesPath to return null for a resource if its resource type is invalid
+        // i.e., String.EMPTY, sling:nonexisting, nt:unstructured, or /var/absolute/path/not/in/resolver/search/path
+        if (sharedPropertiesPath == null) {
+            // if SHARED is requested in this case, return ValueMap.EMPTY
+            if (valueType == SharedComponentProperties.ValueTypes.SHARED) {
+                return ValueMap.EMPTY;
+            } else { // otherwise, pass null for sharedProperties and merge with global and resource
+                return sharedComponentProperties.mergeProperties(globalVmBound, null, resource);
+            }
+        }
+
+        // it is also possible for the shared properties path to differ from the path in bindings for the same
+        // reason that shared properties page path may differ (a wrapped request for a child resource of a
+        // different resource type)
+        final ValueMap sharedVmBound;
+        if (sharedPropertiesPath.equals(bindings.get(SharedComponentProperties.SHARED_PROPERTIES_PATH))) {
```

Comment:
Would it have been a simpler system to use a property like "shared-propertied-<path>" so that we would only need a single property instead of both `SHARED_PROPERTIES_PATH` and `SHARED_PROPERTIES` and then technically could have bindings for multiple resources?

Reply:
I suppose that would work for how we're using the bindings with the SCP API, but it would be a bit unexpected for implicit script variable names, which are taken from the binding property name. I assume they should be legal java identifiers. I'm not sure what would happen with hyphens or slashes.

Reply:
Ok, my only reason for thinking this was in case it somehow enabled better caching of bindings for all the things w/o overwriting each other back and forth for alternating components of different resource types.  I haven't yet run the code, so I dont know if that's applicable or not.  I just see code validating that current bindings are for the current resource, which made me think that it might be possible that bindings are flip flopping, which could be prevented via naming.

Perhaps another way we could resolve would be to put the bindings into a map object, keyed by resource path.

Reply:
It's not so much that they are flip-flopping, because if that happens the caching is still being managed by BVP at the start of the servlet execution for a requested resource. The check for the path is to protect against the case where the cached bindings AREN'T overwritten, when one might assume they should be. I.E. when a Request-adaptable model has a @ChildResourceWithRequest-annotated model member that has its own @SharedValueMapValue annotations. The BVP doesn't have a chance to execute for the request wrapper we inject for the child model, but the parent bindings are still present because we carry along the SlingBindings request attribute in the wrapper class. This means that if we naively check for the presence of a "sharedProperties" binding, it will be for the parent resource, not the current resource. We haven't actually created a new sharedProperties for the child because we are skipping the normal BindingsValuesProvider phase with the wrapped request, so there is no relevant cache value to look up.

Reply:
In that situation, does the changes to the bind values from the child object affect the original request pointing at the parent object?

Reply:
I'm saying those binding values don't change at all for the child resource. The Injector operating on the child doesn't push any changes back to the request bindings. It only decides whether or not to use the bindings based on whether the path bindings match the values that would be expected for child resource. If not, it ignores the bindings altogether.

Reply:
ah k, got it